R/object_name_linters.R

```diff
@@ -1,3 +1,78 @@
+#' @describeIn linters  Check that object names conform to a single naming style.
+#' @param styles One of \Sexpr[stage=render, results=rd]{lintr:::regexes_rd}.
+#' @export
+object_name_linter <- function(styles = "snake_case") {
+  styles <- match.arg(styles, names(style_regexes), several.ok = TRUE)
+
+  function (source_file) {
+    x <- global_xml_parsed_content(source_file)
+    if (is.null(x)) {
+      return()
+    }
+    xpath <- paste0(
+      # Left hand assignments
+      "//expr[SYMBOL or STR_CONST][following-sibling::LEFT_ASSIGN or following-sibling::EQ_ASSIGN]/*",
+
+      # Or
+      " | ",
+
+      # Right hand assignments
+      "//expr[SYMBOL or STR_CONST][preceding-sibling::RIGHT_ASSIGN]/*")
+
+    assignments <- xml2::xml_find_all(x, xpath)
+
+    # Retrieve assigned name
+    nms <- strip_names(
+      as.character(xml2::xml_find_first(assignments, "./text()")))
+
+    generics <- c(declared_s3_generics(x), namespace_imports()$fun, names(.knownS3Generics), .S3PrimitiveGenerics, ls(baseenv()))
+
+    lapply(styles, function(style) {
+      lapply(assignments[!check_style(nms, style, generics)], object_lint2, source_file, paste0("Variable and function names should be in ", style, "."), "object_name_linter")
```

Comment:
The 'assigned-names' that throw lints here, are those that do not match at least one style (from the vector of `styles`). There is a test 'linter accepts a vector of styles' that has a different logic though. It expects lints to be thrown when an assigned-name does not match any of the styles. Could the code logic be reworked to match the test, since it seems unworkable to expect every assigned name to match every one of a set of multiple styles.

Reply:
For example, 
```
lint_msg <- paste0("Variable and function names should be one of ", styles, ".")
style_matches <- lapply(styles, function(x) {
  check_style(nms, x, generics)
})
matches_a_style <- Reduce(`|`, style_matches)
lapply(
  assignments[!matches_a_style],
  object_lint2,
  source_file,
  lint_msg,
  "object_name_linter"
)
```

Reply:
If the lint message is changed to reflect that multiple styles are checked, the expected lint message should be changed in a few tests

Reply:
A modified version of `object_name_linter` that Ive been using locally follows:

```
#' @describeIn linters  Check that object names conform to one of a set of
#'   naming styles.
#' @param styles A subset of
#'   \Sexpr[stage=render, results=rd]{lintr:::regexes_rd}.
#' @export
object_name_linter <- function(styles = "snake_case") {

  .or_string <- function(xs) {
    # returns "<S> or <T>"
    # where <T> is the last string in xs
    # where <S> is a comma-separated string of all entries in xs except <T>
    len <- length(xs)
    if (len <= 1) {
      return(xs)
    }
    comma_sepd_prefix <- paste(xs[-len], collapse = ", ")
    paste(comma_sepd_prefix, "or", xs[len])
  }

  styles <- match.arg(styles, names(style_regexes), several.ok = TRUE)

  lint_msg <- paste0(
    "Variable and function name style should be ", .or_string(styles), "."
  )

  function (source_file) {
    x <- global_xml_parsed_content(source_file)
    if (is.null(x)) {
      return()
    }
    xpath <- paste0(
      # Left hand assignments
      "//expr[SYMBOL or STR_CONST][following-sibling::LEFT_ASSIGN or following-sibling::EQ_ASSIGN]/*",

      # Or
      " | ",

      # Right hand assignments
      "//expr[SYMBOL or STR_CONST][preceding-sibling::RIGHT_ASSIGN]/*")

    assignments <- xml2::xml_find_all(x, xpath)

    # Retrieve assigned name
    nms <- strip_names(
      as.character(xml2::xml_find_first(assignments, "./text()")))

    generics <- c(
      declared_s3_generics(x),
      namespace_imports()$fun,
      names(.knownS3Generics),
      .S3PrimitiveGenerics, ls(baseenv()))

    style_matches <- lapply(styles, function(x) {
      check_style(nms, x, generics)
    })

    matches_a_style <- Reduce(`|`, style_matches)

    lapply(
      assignments[!matches_a_style],
      object_lint2,
      source_file,
      lint_msg,
      "object_name_linter"
    )
  }
}
```

Reply:
For this to match with the existing unit tests, I modified the expected lint-messages in tests "linter returns correct linting" and "linter accepts vector of styles":

```
  # "linter returns correct linting
  msg <- "Variable and function name style should be camelCase."
```

```
  # "linter accepts vector of styles"
  msg <- "Variable and function name style should be camelCase or dotted.case."
```